Premium Blocks: add class to handle premium blocks

retrofox opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to to-test.md in a new commit as part of your PR. -->

This PR adds a ~module~ class to set blocks as premium when they are not registered by Jetpack. It populates the global data object.

Fixes #

#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->
* Add a module to register premium blocks.


#### Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->
no

#### Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

### Simple Site / Free plan.
* Sandbox the site

<img src="https://user-images.githubusercontent.com/77539/89201399-d2134180-d587-11ea-8aab-2859fbf27d7d.png" width="300px" />

* Apply D46955-code in your sandbox in order to register the plan features for the premium blocks: `core/cover`, `core/video`, `core/audio`, and `premium-content/container`.

* Apply these changes in your sandbox too.

* Confirm that you see the blocks are premium in a `free` site

<img src="https://user-images.githubusercontent.com/77539/89202134-e0159200-d588-11ea-8acd-6221d25f5ae6.png" width="500px" />

* Confirm that the blocks are available for in a site with a `paid` plan.

<img src="https://user-images.githubusercontent.com/77539/89202171-ef94db00-d588-11ea-8baf-1be3cbb7baac.png" width="500px" />


#### Proposed changelog entry for your changes:
<!-- Please do not leave this empty. If no changelog entry needed, state as such. -->
* Add a module to register premium blocks.

retrofox commented:
> Part of me wonders if we should have that file within `extensions/blocks` to keep as much block-related code as possible in that part of the codebase; it may make things easier to understand for new contributors.

Sounds great to me.

> Could you address the feedback from Danger bot so the tests can pass?

if I move the code to `extensions/blocks` then we won't need to address the Danger bot feedback, correct?

jeherve commented:
> if I move the code to extensions/blocks then we won't need to address the Danger bot feedback, correct?

Right, exactly, because the whole `extensions/` directory is already in there:
https://github.com/Automattic/jetpack/blob/7aed4db3d5d49149fdcbbf8eb5666b74d3adffde/bin/phpcs-requirelist.js#L17

jeherve commented on modules/premium-blocks.php:
```diff
@@ -0,0 +1,89 @@
+ * @package Jetpack
+ */
+
+class Premium_Blocks {
```
I would recommend either prefixing that class name or putting it in the `\Automattic\Jetpack\Extensions\` namespace, to avoid any conflicts with third-party plugins.

retrofox commented:
> Right, exactly, because the whole `extensions/` directory is already in there:

We've created the `extensions/share/premium-blocks/` folder in the [Premium Blocks](https://github.com/Automattic/jetpack/pull/16611) approach, so I'm going to add this class right there. Happy to change the path is you consider there is a better location.

retrofox commented:
Thanks @jeherve for your help.
I've updated the code according to feedback. Also created a D47508-code. Not sure how to proceed with the sync process.

jeherve commented:
cc @pablinos who worked on #14426 and may have some good feedback on this.

retrofox commented:
> Let's add a few more comments to properties and methods before this gets in. I'm not sure those properties need to be static?

I've added some comments but I think you've addressed most of them

> I branched off from your PR and created a commit with my suggestions since they were quite extensive but low priority: [1ffa52d](https://github.com/Automattic/jetpack/commit/1ffa52dff24c029fe64cd64202ae65932d0498c6)
> 
> Feel free to cherry-pick or ignore.

✅ 

Thanks for your review and feedback (included the free commit! 🤞 )

jeherve commented:
Oh, I see you're already working on consolidating things in #16611. :)

retrofox commented on extensions/shared/premium-blocks/class-premium-blocks.php:
```diff
@@ -0,0 +1,88 @@
+	 *
+	 * @var string
+	 */
+	public $required_plan = 'jetpack_premium';
```
Yeah, sorry. Let me remove it.

retrofox commented:
> Oh, I see you're already working on consolidating things in #16611. :)

yes... that's right. However, the scope of this PR is to set/register some blocks as premium. #16611 will take over of the client-side implementation. Testing both together is a nice experience as well :-D.

retrofox commented:
> If we extend this later on to support more plans, how do you think we should approach things? Should we then load a new `extensions/shared/business-blocks` for example?

I don't think so. If I'm not wrong, we can define the scope of the paid site on the dotcom side? Keep in mind we need to apply the changes defined here D46955-code to make the approach works as expected.

> I am asking to try to make things as straightforward as possible for new contributors. We currently take the "`plan_check`" approach ([example](https://github.com/Automattic/jetpack/blob/57fc4e31c38134885be595e0b3a02d359d7a5065/extensions/blocks/send-a-message/send-a-message.php#L28)) for some of our paid blocks, ...

That's right, but afaik we use `jetpack_register_block()` to register new blocks in Jetpack. This PR tries to deal with blocks not registered by Jetpack but for other plugins (core, ETP, etc.)

a different approach for [our VideoPress extension](https://github.com/Automattic/jetpack/blob/9c0bcafb1743fce9f14861637f705f4c9ce115c6/modules/videopress/class.videopress-gutenberg.php), and [for the Payments block](https://github.com/Automattic/jetpack/blob/5aa736086186f5e00fc36110583da4d61e2b0a9f/modules/memberships/class-jetpack-memberships.php#L361-L377). I worry this may be a bit confusing for new contributors, and an additional method may add more confusion.

I think we can simplify videopress core/video with this approach. Let me take a look at payments block too.


> Maybe updating our documentation would help?

Yes, let me update it.

> https://github.com/Automattic/jetpack/blob/5a4ec0da4f32874708b4f7014690aa458ceca883/extensions/README.md#L157

retrofox commented:
> If we extend this later on to support more plans, how do you think we should approach things? Should we then load a new `extensions/shared/business-blocks` for example?

oh, should we replace `premium` by `paid`?

obenland commented:
Is Jetpack the best place for this class? It looks like it's non-operational outside of simple sites, should we maybe move this to dotcom?

retrofox commented:
> Is Jetpack the best place for this class? It looks like it's non-operational outside of simple sites, should we maybe move this to dotcom?

This approach requires to implement changes on both sides. `D46955-code` does the dotcom side.
However, we need to set these blocks are premium too in Jetpack. I prefer saying _set_ instead of _register_ because we aren't registering these blocks. They are already registered by other plugins (core, ETP, or whatever).

The idea is to populate the Jetpack data global object to finally handle the pieces of the Upgrade feature: upgrade banner in the editor canvas, upgrade banner in the block inspector, extending icons, cover/core and core/audio as special cases since are usable with a not paid plan.
You can check the whole implementation applying https://github.com/Automattic/jetpack/pull/16611

pablinos commented:
I've been pondering on this a little overnight, and I'm still not sure of the best solution, but some things I'm thinking we should look at are:

- Making sure lists of features are co-located in the plan classes `class-jetpack-plan.php` and the `Store_Product_list` on the wpcom side. Plans should have features associated with them, and we need to make sure that we don't need to update another list somewhere else when we add a new block. 
- Related to the above, we should also make sure the business logic of which features are available is in these classes and passed to the client via the API etc. We should be avoiding using `isSimpleSite` etc. on the client-side.
- The Premium Content plugin/block should be brought into Jetpack as a beta block or similar. I assume it was in FSE to ease rapid development, but if it's coming out of that phase, I think it makes sense to add it to Jetpack
- The problem with the other blocks is that we're extending core functionality so the block name and the Jetpack module/extension aren't inline. We also want to have custom logic for different blocks. I'd propose handling this in `register_jetpack_block`. Module and extension names could be passed in as further options, along with maybe a callback or other way to perform a test. This does assume that we call `register_jetpack_block` for all the core blocks we extend. This is the case for `core/video` but I don't know about the others. Perhaps we could make it the case?
- The cover block is a bit of a problem, as it becomes premium based on its usage. How we're handling this is ok at the moment, but it breaks the principles above. For speed, we should probably treat this as a special case, and revisit it later.

That last point seems important to me. There's probably a need to get this working quickly and so perhaps we go with what we've got but commit to refactoring in a maintenance cycle.

retrofox commented:
> Making sure lists of features are co-located in the plan classes `class-jetpack-plan.php` and the `Store_Product_list` on the wpcom side. Plans should have features associated with them, and we need to make sure that we don't need to update another list somewhere else when we add a new block.

The point is we don't have a premium blocks list in Jetpack so far because hadn't to deal whit this issue: Set blocks not registered by Jetpack as Premium blocks.
But in some way (a different one) that's what we are doing right now. `core/video` block has its own code to handle the upgrade nudge, and `core/cover` follows almost the same logic. Basically, hardcoding a premium blocks list in Jetpack, on the client-side.
Also, set explicitly a block as premium is needed when we register the block, passing the `'plan_check' => true`.

> Related to the above, we should also make sure the business logic of which features are available is in these classes and passed to the client via the API etc. We should be avoiding using isSimpleSite etc. on the client-side.

Fair enough. Agree.
Side question. Is it possible to set blocks as premium for not-simple sites? I guess that could happen in Atomic sites, but not sure since an Atomic site is not free by definition.

> The Premium Content plugin/block should be brought into Jetpack as a beta block or similar. I assume it was in FSE to ease rapid development, but if it's coming out of that phase, I think it makes sense to add it to Jetpack

I think this an exception. Something similar happened with `donations`. Now is part of Jetpack and we don't have to take care of this block because, since Jetpack registers it, it deals with the business logic. But this isn't scalable. We don't know which blocks will be treated as premium in the short term, even core blocks.

> The problem with the other blocks is that we're extending core functionality so the block name and the Jetpack module/extension aren't inline.

That's an issue inherited from the complexity of the implementation. Why not just set/expose a premium blocks list from? It's really complicated to realize that `videopress`feature is tied to the `core/video` block.

> We also want to have custom logic for different blocks. I'd propose handling this in register_jetpack_block. Module and extension names could be passed in as further options, along with maybe a callback or other way to perform a test.

Could you show a simple example of this could look? It sounds kinda confusing to me. We are using the register function to register blocks that are already registered by other plugins. I'm not saying it's wrong just trying to understand the approach.

> This does assume that we call register_jetpack_block for all the core blocks we extend.

Yes, agree.

> This is the case for core/video but I don't know about the others. Perhaps we could make it the case?

`core/video`, `core/audio`, `core/cover` and `premium-content/container` so far. 

> The cover block is a bit of a problem, as it becomes premium based on its usage. How we're handling this is ok at the moment, but it breaks the principles above. For speed, we should probably treat this as a special case, and revisit it later.

First, this isn't the only one block with this issue. `core/audio` has a similar issue and even `core/video` too. Are blocks that can change their behavior and/or functionality depending on the current plan of the site. 
Second, I'm afraid the blocks list could increase since, IMHO, I see here a business opportunity too. The same concept of free/paid site could be applied to blocks.

The harder part of this is, in general, we can't define how some blocks should act according to the plan where it belongs. Are not the same restrictions for the `core/cover` that the `core/audio` or `core/video`. However, we can generalize and implement common behaviors, for instance, the media placeholder.

> That last point seems important to me. There's probably a need to get this working quickly and so perhaps we go with what we've got but commit to refactoring in a maintenance cycle.

Yesterday [I did a refactoring](https://github.com/Automattic/jetpack/pull/16717) in order to make the upgrade banner work for `core/audio`. 

Among many changes, I've defined a blocks list that can be usable with a free plan, and also added some data used to determine which features we should handle.

```es6
const usableBlockWithFreePlan = [
	{
		name: 'core/cover',
		mediaPlaceholder: true,
		mediaReplaceFlow: true,
		fileType: 'video',
	},
	{
		name: 'core/audio',
		mediaPlaceholder: true,
		fileType: 'audio',
	},
];
```

That results in it controls the MediaPlacholder and MediaReplaceButton for the `core/cover` and MediaPlaceholder for `core/audio`.

retrofox commented:
> Making sure lists of features are co-located in the plan classes `class-jetpack-plan.php` ...

Sorry about this question: this is only for Jetpack sites, right? That class is not loaded in dotcom side.

pablinos commented:
Thanks for the further explanation @retrofox and sorry for all the back and forth we're having on this. It does seem like this is something we should get right, but I have a habit of labouring points a bit too much 😄 I'm not totally against this solution as it gets the job done, but something isn't sitting right and I'm struggling to put my finger on it.

> The point is we don't have a premium blocks list in Jetpack so far because hadn't to deal whit this issue: Set blocks not registered by Jetpack as Premium blocks.

We shouldn't be making blocks premium that aren't registered by Jetpack. If it's a core block then we're extending functionality and that extended block can be registered through Jetpack. That's another big argument for bringing the premium content block over from FSE.

There's been a lot of discussion about where block development should happen, and the consensus was that this repository is the place, so any premium blocks should be able to be registered using the Jetpack mechanisms.

> Also, set explicitly a block as premium is needed when we register the block, passing the 'plan_check' => true.

This is a good point, and we probably shouldn't need that in the future. It came about as I wanted to selectively turn this plan gating functionality on. You'll notice that there's a lot of free blocks not listed in the free plan.

> Is it possible to set blocks as premium for not-simple sites? I guess that could happen in Atomic sites, but not sure since an Atomic site is not free by definition.

We've left Atomic as having no restrictions. The idea is that they're on a Business a plan. I realise this is a bit of hack in itself, but it was the simplest option at the time. So no, not currently, but we could implement Atomic logic based on the Business plan as Jetpack runs on Atomic. We sholudn't be thinking about Atomic and Simple sites, more what plan does a site have?

> Why not just set/expose a premium blocks list from?

We can, but this should be a list of features associated with a plan

> core/video, core/audio, core/cover and premium-content/container so far.

Yes, but are the extensions to these plugins utilising `register_jetpack_block`. Not for premium content, and I don't think `core/cover` but could they be made to, so that we're using the same pattern?

> Could you show a simple example of this could look? 

I need to give it some more thought, but I would hope we could define a way of structuring the feature information in the plans classes, so when the block is registered (or the availability set) all the checks could be done.

> It sounds kinda confusing to me. We are using the register function to register blocks that are already registered by other plugins.

We shouldn't be registering blocks that are already registered by other plugins. If they're core blocks then we're extending them and so we're registering our premium versions. The premium content feature is the odd one here, but that further builds the case for bringing it into Jetpack 😄 

> First, this isn't the only one block with this issue. core/audio has a similar issue and even core/video too

If I understand it correctly, the cover block is different, as it can be partly used without a plan as long as you're not using a video background, whereas the `core/audio` and `core/video` blocks are all or nothing, based on your plan. The mechanisms we have in place, and we've been discussing have been about blocks being wholly available on a plan or not. I suppose we could get around this with an Advanced Cover block or something, but the idea of a block becoming premium as a person swaps from using an image to a video is tricky to deal with in a generic way. It's also an odd UX in my opinion.

> The harder part of this is, in general, we can't define how some blocks should act according to the plan where it belongs. Are not the same restrictions for the core/cover that the core/audio or core/video. However, we can generalize and implement common behaviors, for instance, the media placeholder.

I'm not sure what you mean by this, but it sounds similar to what I'm describing above, where we want to gate features of a block rather than the whole block itself. Dealing with that in a generic way seems like a tricky problem to solve.

> That results in it controls the MediaPlacholder and MediaReplaceButton for the core/cover and MediaPlaceholder for core/audio

As I say, this is a solution that gets the job done, but it's another list in the code, and seems tied to these media blocks. 

I'm not against this approach, but I think we will need to change it down the road. Can we get this configuration information in one place, and register the blocks via Jetpack? If so, that would go a long way towards solving the problem.

pablinos commented:
> Sorry about this question: this is only for Jetpack sites, right? That class is not loaded in dotcom side.

Yes, that's right, there's the Store_Product_List (I think that's what it's called) on the dotcom side.

retrofox commented:
> 😄 I'm not totally against this solution as it gets the job done, but something isn't sitting right and I'm struggling to put my finger on it.

Please don't feel bad giving your feedback. What we want to do is trying to achieve the best approach for this purpose. Let's change/discard what we need! ✋ 

> We sholudn't be thinking about Atomic and Simple sites, more what plan does a site have?

💯 

>> Why not just set/expose a premium blocks list from?
>
> We can, but this should be a list of features associated with a plan

Exactly, in a similar way that the features are exposed and pulled fro Jetpack to define if a block scope in terms of the site plan, for instance `opentable`, `send-a-message`, etc. The difference here is we don't know, in jetpack side, what are the core, FSE, whatever blocks that need to be treated as premium blocks. And we need to know it because the Upgrade stuff lives in Jetpack: showing an Upgrade banner when it's required.

> Yes, but are the extensions to these plugins utilising register_jetpack_block. Not for premium content, and I don't think core/cover but could they be made to, so that we're using the same pattern?

Not sure. We can re-use some helper functions, etc. but the problem is not exactly the same.

> ...but I would hope we could define a way of structuring the feature information in the plans classes, so when the block is registered (or the availability set) all the checks could be done.

You mean to the Jetpack sites, but what about the dotcom? That's just what this PR does. Set the ability of some core, FSE blocks are premium

> We shouldn't be registering blocks that are already registered by other plugins. If they're core blocks then we're extending them and so we're registering our premium versions. 

Again, that's what this PR tries to achieve. :-D

> If I understand it correctly, the cover block is different, as it can be partly used without a plan as long as you're not using a video background, whereas the core/audio and core/video blocks are all or nothing, based on your plan.

I guess you can set an audio or video file using an URL. And it should be doable beyond what's the plan that the site has.

retrofox commented:
I've updated the PR with the following significant changes:

* Do not create a static list for these blocks. Instead, it pulls the list from dotcom class calling the `get_paid_blocks_list` method, added on D46955-code
* Improve documentation

With these changes, we don't have to take care of settings these blocks in the Jetpack side, and it makes sense to me since these blocks are not registered by Jetpack.

You can test the whole implementation by applying these changes, #16611 and D46955-code.

retrofox commented:
Ive 

> Is Jetpack the best place for this class? It looks like it's non-operational outside of simple sites, should we maybe move this to dotcom?

I've moved this implementation to D46955-code. Props to @obenland to suggest it (twice). 
Thanks folks for your help. Closing the PR.
